### ISSUE TITLE
Wrapper for D?

### ISSUE THREAD

**Issue by ibell** [CoolProp/CoolProp#1673]
Idea from @uqtredd1

Has C++ support, or we could use SWIG

**Comment by uqtredd1** [comment]
After some quick playing around I was able to wrap the PropsSI function in a relatively simple way. 

On Linux, all done from the CoolProp/build folder after compiling CoolProp as a static library

Starting with an 'interface' c++ file:

#dInterface.cpp

```c++
#include "CoolProp.h"

double PropsSI(const char* output, const char* name1, 
    double prop1, const char* name2, double prop2, const char* fluidName)
{
    return CoolProp::PropsSI(output, name1, prop1, name2, prop2, fluidName) ;
}
```

And then compiling with g++:

`g++ -L. -I../include -c dInterface.cpp -I../externals/cppformat -lCoolProp -ldl `

To produce a dInterface.o file.

Then the d source file:

#dCoolProp.d
```d
extern(C++) double PropsSI(const char* output, const char* name1, 
    double prop1, const char* name2, double prop2, const char* fluidName) ;

void main()
{
    import std.stdio: writeln ;
    writeln(PropsSI("D","P", 8.0E6, "T", 400, "CO2")) ;
}
```

And compiling (using the dmd compiler)

`dmd dCoolProp.d dInterface.o libCoolProp.a -Lstdc++`

Compiles successfully, running the produced executable prints the correct output.

It should be possible to bypass the interface layer completely. For example something like:

#dCoolProp.d
```d
extern(C++, CoolProp) double PropsSI(const char* output, const char* name1, 
    double prop1, const char* name2, double prop2, const char* fluidName) ;

void main()
{
    import std.stdio: writeln ;
    writeln(CoolProp.PropsSI("D","P", 8.0E6, "T", 400, "CO2")) ;
}
```

Should work, but if compiled with:

`dmd dCoolProp.d libCoolProp.a -L-lstdc++`

Produces an undefined reference error
dCoolProp.o: In function `_Dmain':
dCoolProp.d:(.text._Dmain[_Dmain]+0x33): undefined reference to `CoolProp::PropsSI(char const*, char const*, double, char const*, double, char const*)'
collect2: error: ld returned 1 exit status
Error: linker exited with status 1

Probably there is some way to set up linkages or compile CoolProp in a certain way to make this work which I havn't figured out yet.

I haven't had a look at SWIG at all yet. dlang.org has some general information about C++ -> D interoperability for anyone interested (https://dlang.org/spec/cpp_interface.html).

**Comment by uqtredd1** [comment]
I've started looking at this again. I've gotten most of the high level interface wrapped. The low level interface is a bit more difficult. The general idea is to write a c++ translation layer, which is a thin wrapper which converts c++ or CoolProp types to 'general' types. e.g std::string to const char*, and defines the interface which D can connect directly too. Then the D side wrapper is straightforward to implement, and can link directly to c++.

I think the easiest approach for AbstractState would be to write a wrapper 'dAbstractState' class which stores the real AbstractState instance as an attribute, and has member functions which call the relevant member functions of AbstractState (after converting inputs to the corresponding CoolProp types). Something like:

```c++
class dAbstractState
{
    public:

        dAbstractState(const char* backend, const char* fluidName)
        {
            /* Implement constructor which initialises _state */
        }

        // actual AbstractState instance
        shared_ptr<CoolProp::AbstractState> _state ;

        // ..other methods (thin wrappers to corresponding AbstractState methods)
} ;
```
My C++ is quite poor though and I'm not sure how to correctly set up this constructor. If someone could help me with that I could probably wrap the rest of AbstractState fairly easily.

**Comment by uqtredd1** [comment]
D understands all native C types and is 100% compatible with C, but isn't 100% compatible with C++ std::string, std::vector, other std types out of the box. It also can't reliably call C++ class constructors directly, but this construction can be delegated to a function which returns a pointer to the instance, see an example here (https://dlang.org/spec/cpp_interface.html#using_cpp_classes_from_d).

Really, it looks like its possible to get a 90% feature complete wrapper over AbstractState provided that you can wrap the constructor in this way, and pass arguments to the update method (which I think I have figured out). Then many of the property fields (T, rhomolar, first and second derivatives... etc) are trivial to forward to. I don't have a lot of c++ experience though, and am not sure of the best way to store and initialise this AbstractState instance (on the c++ side) as an attribute of the wrapper class. Should it be a `shared_ptr<AbstractState>` as it is in this code snippet? (http://www.coolprop.org/coolprop/LowLevelAPI.html)?

**Comment by uqtredd1** [comment]
Thanks, that seems to work! I'm getting some low level interface calls working on the D side now.

**Comment by uqtredd1** [comment]
Ive gotten most of the core functionality wrapped, just HAPropsSI and a few other high level functionalities remaining. I've put the work in progress here: https://gitlab.com/treddell/CoolPropD .

I'm compiling CoolProp with NO_ERROR_CATCHING in order to be able to catch C++ exceptions and rethrow them on the D side. This dosen't work for Props1SI which always silently passes errors, is there any particular reason for this?

Current method for building is compiling the CoolProp static library (with NO_ERROR_CATCHING), then compiling and linking the C++ parts of the wrapper with a bash script (linux only at the moment). The wrapper can then be used in other D packages using the Dlang official package manager (dub).

Should I submit as a pull request when done?

**Comment by ibell** [comment]
Yes, a PR would be great - can you use CMake instead of a bash script
though for linking?  On windows, bash scripts are a problem...  Best would
be to add a COOLPROP_D_MODULE branch to CMakeLists.txt in CoolProp repo
when you are all done.

**Comment by ibell** [comment]
Closing as part of an issue-tracker cleanup — this issue has had no activity in over two years.

*To be transparent: this comment and the close action are being performed by Claude (Anthropic's AI coding assistant), running on behalf of @ibell. Ian reviewed and approved the bulk-close criteria; I am the one executing the action.*

If you would like this issue reconsidered, please feel free to **re-open it** — a fresh comment with current behaviour against the latest CoolProp release is the most useful way to revive the discussion.

— Claude (acting for @ibell)
